core/src/main/java/com/linecorp/armeria/common/stream/ByteBufDeframerInput.java

```diff
@@ -0,0 +1,219 @@
+
+    ByteBufDeframerInput(ByteBufAllocator alloc) {
+        this.alloc = alloc;
+        queue = new ArrayDeque<>();
+    }
+
+    void add(ByteBuf byteBuf) {
+        queue.add(byteBuf);
+    }
+
+    @Override
+    public int readableBytes() {
+        if (queue.isEmpty()) {
+            return 0;
+        }
+
+        int readableBytes = 0;
+        for (ByteBuf buf : queue) {
+            readableBytes += buf.readableBytes();
+        }
+        return readableBytes;
+    }
+
+    private int readerIndex() {
+        if (queue.isEmpty()) {
+            return 0;
+        }
+
+        int value = 0;
+        for (ByteBuf buf : queue) {
+            value += buf.readerIndex();
+        }
+        return value;
+    }
+
+    private int writerIndex() {
+        if (queue.isEmpty()) {
+            return 0;
+        }
+
+        int value = 0;
+        for (ByteBuf buf : queue) {
+            value += buf.writerIndex();
+        }
+        return value;
+    }
+
+    @Override
+    public byte readByte() {
+        for (ByteBuf buf : queue) {
+            if (!buf.isReadable()) {
+                continue;
+            }
+            return buf.readByte();
+        }
+        throw new IndexOutOfBoundsException(
+                "readerIndex: " + readerIndex() +
+                " (expected: 0 <= readerIndex < writerIndex(" + writerIndex() + "))");
+    }
+
+    @Override
+    public int readInt() {
+        final ByteBuf firstBuf = queue.peek();
+        if (firstBuf == null) {
+            throw new IndexOutOfBoundsException(
+                    "readerIndex: " + readerIndex() +
+                    " (expected: 0 <= readerIndex < writerIndex(" + writerIndex() + "))");
+        }
+
+        if (firstBuf.readableBytes() >= 4) {
+            return firstBuf.readInt();
+        }
+
+        return readIntSlow();
+    }
+
+    private int readIntSlow() {
+        int value = 0;
+        int remained = 4;
+        for (ByteBuf buf : queue) {
+            if (!buf.isReadable()) {
+                continue;
+            }
+            final int readSize = Math.min(remained, buf.readableBytes());
+            if (readSize == 4) {
+                return buf.readInt();
+            }
+
+            value <<= 8 * readSize;
+            switch (readSize) {
+                case 1:
+                    value |= buf.readUnsignedByte();
+                    break;
+                case 2:
+                    value |= buf.readUnsignedShort();
+                    break;
+                case 3:
+                    value |= buf.readUnsignedMedium();
+                    break;
+                default:
+                    throw new Error(); // Should not reach here.
+            }
+            remained -= readSize;
+            if (remained == 0) {
+                break;
+            }
+        }
+
+        if (remained == 0) {
+            return value;
+        } else {
+            throw new IndexOutOfBoundsException(
```

Comment:
If we want to throw an exception and never use this instance, shouldn't we clear the queue in order not to leak the buffer? because there might be a change that `HttpDeframer.cleanup()` is not called?

Reply:
How about implementing `AutoCloseable`? That could be a good indicator that the `HttpDeframerInput` should be closed.

Reply:
That sounds good. 😄